cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -896,48 +896,106 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
-            let (f32_op, f64_op) = match op {
-                Opcode::Fadd => (SseOpcode::Addss, SseOpcode::Addsd),
-                Opcode::Fsub => (SseOpcode::Subss, SseOpcode::Subsd),
-                Opcode::Fmul => (SseOpcode::Mulss, SseOpcode::Mulsd),
-                Opcode::Fdiv => (SseOpcode::Divss, SseOpcode::Divsd),
-                _ => unreachable!(),
+            // instruction will return the second operand if either operand is a NaN.
+            let sse_op = match ty {
+                types::F32 => match op {
+                    Opcode::Fadd => SseOpcode::Addss,
+                    Opcode::Fsub => SseOpcode::Subss,
+                    Opcode::Fmul => SseOpcode::Mulss,
+                    Opcode::Fdiv => SseOpcode::Divss,
+                    _ => unreachable!(),
+                },
+                types::F64 => match op {
+                    Opcode::Fadd => SseOpcode::Addsd,
+                    Opcode::Fsub => SseOpcode::Subsd,
+                    Opcode::Fmul => SseOpcode::Mulsd,
+                    Opcode::Fdiv => SseOpcode::Divsd,
+                    _ => unreachable!(),
+                },
+                types::F32X4 => match op {
+                    Opcode::Fadd => SseOpcode::Addps,
+                    Opcode::Fsub => SseOpcode::Subps,
+                    Opcode::Fmul => SseOpcode::Mulps,
+                    Opcode::Fdiv => SseOpcode::Divps,
+                    _ => unreachable!(),
+                },
+                types::F64X2 => match op {
+                    Opcode::Fadd => SseOpcode::Addpd,
+                    Opcode::Fsub => SseOpcode::Subpd,
+                    Opcode::Fmul => SseOpcode::Mulpd,
+                    Opcode::Fdiv => SseOpcode::Divpd,
+                    _ => unreachable!(),
+                },
+                _ => panic!(
+                    "invalid type: expected one of [F32, F64, F32X4, F64X2], found {}",
+                    ty
+                ),
             };
+            ctx.emit(Inst::xmm_rm_r(sse_op, rhs, dst));
+        }
 
-            let is_64 = flt_ty_is_64(ty.unwrap());
+        Opcode::Fneg => {
+            let src = input_to_reg_mem(ctx, inputs[0]);
+            let dst = output_to_reg(ctx, outputs[0]);
+            let ty = ty.unwrap();
+
+            // Move the `lhs` to the same register as `dst`; this may not emit an actual move
+            // but ensures that the registers are the same to match x86's read-write operand
+            // encoding.
+            match src {
+                RegMem::Reg { reg } => ctx.emit(Inst::gen_move(dst, reg, ty)),
+                RegMem::Mem { .. } => unimplemented!(),
+            }
 
-            let mov_op = if is_64 {
-                SseOpcode::Movsd
-            } else {
-                SseOpcode::Movss
-            };
-            ctx.emit(Inst::xmm_mov(mov_op, lhs, dst, None));
+            // Generate an all 1s constant in an XMM register.
+            let tmp = ctx.alloc_tmp(RegClass::V128, ty);
```

Comment:
@cfallin I had to switch over to this version of `fneg` due to negating signed 0s (previously I had the "subtract the value from 0.0 version") but I believe that using this temporary register is causing this to fail again in the register allocator?

`thread '<unnamed>' panicked at 'register allocation: Analysis(EntryLiveinValues)', cranelift/codegen/src/machinst/compile.rs:73:9`

Reply:
@cfallin proposed to add a `Inst::InputInvariant` (or perhaps `Input::InputClobber`) that would allow us to use `PXOR`, `CMPEQPS`, etc., to generate certain constants. That new `Inst` would have a single `Writable<Reg>` and it's `get_reg` would contain one def and no uses.

Reply:
@bnjbvr and I discussed this (the use of `PXOR r,r`, `CMPEQPS r,r` etc) to generate constants in registers, and the effect on the regalloc.  All that is needed to keep regalloc happy is for `get_regs` to spot such cases (where the two registers are the same) and report a single `def` for the register, instead of the usual `use+mod` pair.  No need for any `Inst::InputInvariant` or `Input::InputClobber` or any other form of hinting.

Reply:
+1 to @julian-seward1's suggestion to reuse the existing instructions for this purpose.

Reply:
Ok, makes sense.

Reply:
One concern I'd have in using the same instruction format for both e.g. `PXOR` and `PAND` (if that's a thing? let's say it exists as a straw-instruction if not), with the `get_regs` logic to special-case the same-register case, is that the input-invariant behavior really does depend on the opcode. So `PXOR r1, r1` zeroes `r1`, and doesn't use its initial value; but `PAND r1, r1` leaves `r1`'s final value equal to its initial value (so uses `r1`).

Using a `mod` for the first reg would solve this issue (`PAND` would still be correct), but then doesn't give us the behavior we want for `PXOR`, because `r1` is still (incorrectly, or rather imprecisely) seen as an input, even though the instruction semantically does `r1 := 0` no matter what.

So I think we either need variants of `Inst`, or else we need a flag inside of this variant, or else we need logic in `get_regs` to look at the opcode, to know the difference between ops where `a OP b = constant` and `a OP b = f(a, b)`. I'd prefer the first (it feels cleanest), but any is fine. Thoughts?

Reply:
(Actually, I'm warming up to the third option now; basically we want a helper `op_on_self_yields_constant` or somesuch, and call that in `get_regs`, which is just a match yielding a bool, which is simple enough.)

Reply:
For me, the simplest and most obvious solution is also the third.  In short we want `get_regs` to always tell us the exact truth, and if it requires looking at the opcode and/or operands, so be it.  To allow a situation where it doesn't tell us the exact truth strikes me as setting a dangerous precedent.

I wouldn't even bother with a helper -- these cases are sufficiently rare to make that unnecessary.  I'd just write something like
```
if opcode == XOR && src.isReg() && src.Reg() == dst { // special case }
```
It's one extra line (more or less).